components/support/sync15-traits/src/bridged_engine.rs

```diff
@@ -0,0 +1,199 @@
+    /// times per sync, once for each batch. Implementations can use the
+    /// signal to check if the operation was aborted, and cancel any
+    /// pending work.
+    fn store_incoming(
+        &self,
+        incoming_cleartexts: &[String],
+        signal: &dyn Interruptee,
+    ) -> Result<(), Self::Error>;
+
+    /// Applies all staged records, reconciling changes on both sides and
+    /// resolving conflicts. Returns a list of records to upload.
+    fn apply(&self, signal: &dyn Interruptee) -> Result<BridgedSyncResults, Self::Error>;
+
+    /// Indicates that the given record IDs were uploaded successfully to the
+    /// server. This is called multiple times per sync, once for each batch
+    /// upload.
+    fn set_uploaded(
+        &self,
+        server_modified_millis: i64,
+        ids: &[String],
+        signal: &dyn Interruptee,
+    ) -> Result<(), Self::Error>;
+
+    /// Indicates that all records have been uploaded. At this point, any record
+    /// IDs marked for upload that haven't been passed to `set_uploaded`, can be
+    /// assumed to have failed: for example, because the server rejected a record
+    /// with an invalid TTL or sort index.
+    fn sync_finished(&self, signal: &dyn Interruptee) -> Result<(), Self::Error>;
+
+    /// Resets all local Sync state, including any change flags, mirrors, and
+    /// the last sync time, such that the next sync is treated as a first sync
+    /// with all new local data. Does not erase any local user data.
+    fn reset(&self) -> Result<(), Self::Error>;
+
+    /// Erases all local user data for this collection, and any Sync metadata.
+    /// This method is destructive, and unused for most collections.
+    fn wipe(&self) -> Result<(), Self::Error>;
+
+    /// Tears down the engine. The opposite of `initialize`, `finalize` is
+    /// called when an engine is disabled, or otherwise no longer needed. The
+    /// default implementation does nothing.
+    fn finalize(&self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct BridgedSyncResults {
+    /// List of records
+    pub records: Vec<String>,
+    /// None indicates we aren't reporting this information
+    pub num_reconciled: Option<usize>,
+}
+
+impl BridgedSyncResults {
+    pub fn new(records: Vec<String>, num_reconciled: impl Into<Option<usize>>) -> Self {
+        Self {
+            records,
+            num_reconciled: num_reconciled.into(),
+        }
+    }
+}
+
+// Shorthand for engines that don't care.
+impl From<Vec<String>> for BridgedSyncResults {
+    fn from(records: Vec<String>) -> Self {
+        Self {
+            records,
+            num_reconciled: None,
+        }
+    }
+}
+
+/// A blanket implementation of `BridgedEngine` for any `Mutex<BridgedEngine>`.
+/// This is provided for convenience, since we expect most bridges to hold
+/// their engines in an `Arc<Mutex<impl BridgedEngine>>`.
+impl<E> BridgedEngine for Mutex<E>
```

Comment:
I'm wondering if to trash this now...in theory, if we make all our `BridgedEngineImpl`s `Send + Sync` (if it holds a `Mutex<StorageDb>` internally), the XPCOM newtype won't need to wrap it in a `Mutex`—it could just hold an `Arc<BridgedEngineImpl>` instead of an `Arc<Mutex<BridgedEngineImpl>>`. But we can do that in a follow-up, too.

Reply:
I'll leave this for a follow-up